cpp/dolfinx/fem/FiniteElement.cpp

```diff
@@ -79,62 +90,78 @@ FiniteElement::FiniteElement(const ufc_finite_element& ufc_element)
+}
 //-----------------------------------------------------------------------------
-int FiniteElement::block_size() const { return _block_size; }
+int FiniteElement::block_size() const noexcept { return _bs; }
 //-----------------------------------------------------------------------------
 int FiniteElement::value_dimension(int i) const
 {
   if (i >= (int)_value_dimension.size())
     return 1;
-  return _value_dimension.at(i);
+  else
+    return _value_dimension.at(i);
 }
 //-----------------------------------------------------------------------------
-std::string FiniteElement::family() const { return _family; }
+std::string FiniteElement::family() const noexcept { return _family; }
 //-----------------------------------------------------------------------------
 void FiniteElement::evaluate_reference_basis(
     Eigen::Tensor<double, 3, Eigen::RowMajor>& reference_values,
     const Eigen::Ref<const Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic,
                                         Eigen::RowMajor>>& X) const
 {
-  assert(_evaluate_reference_basis);
-  const int num_points = X.rows();
-  int ret = _evaluate_reference_basis(reference_values.data(), num_points,
-                                      X.data());
-  if (ret == -1)
-  {
-    throw std::runtime_error("Generated code returned error "
-                             "in evaluate_reference_basis");
-  }
+  Eigen::ArrayXXd basix_data = basix::tabulate(_basix_element_handle, 0, X)[0];
+
+  const int scalar_reference_value_size = _reference_value_size / _bs;
+
+  assert(basix_data.cols() % scalar_reference_value_size == 0);
+  const int scalar_dofs = basix_data.cols() / scalar_reference_value_size;
+
+  assert(reference_values.dimension(0) == X.rows());
+  assert(reference_values.dimension(1) == scalar_dofs);
+  assert(reference_values.dimension(2) == scalar_reference_value_size);
+
+  assert(basix_data.rows() == X.rows());
+
+  for (int p = 0; p < X.rows(); ++p)
+    for (int d = 0; d < scalar_dofs; ++d)
+      for (int v = 0; v < scalar_reference_value_size; ++v)
+        reference_values(p, d, v) = basix_data(p, d + scalar_dofs * v);
 }
 //-----------------------------------------------------------------------------
 void FiniteElement::evaluate_reference_basis_derivatives(
     Eigen::Tensor<double, 4, Eigen::RowMajor>& reference_values, int order,
     const Eigen::Ref<const Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic,
                                         Eigen::RowMajor>>& X) const
 {
-  assert(_evaluate_reference_basis_derivatives);
-  const int num_points = X.rows();
-  int ret = _evaluate_reference_basis_derivatives(reference_values.data(),
-                                                  order, num_points, X.data());
-  if (ret == -1)
-  {
-    throw std::runtime_error("Generated code returned error "
-                             "in evaluate_reference_basis_derivatives");
-  }
+  assert(order == 1); // TODO: fix this for order > 1
```

Comment:
Is it particularly hard to generalize this? Seems like the old version supports arbitrary order.

Reply:
I've made issue #1339